tests/integration/cmd_app_test.go

```diff
@@ -27,102 +23,128 @@ var _ = Describe("odo app command tests", func() {
-
-			appListOutput := helper.Cmd("odo", "app", "list", "--project", commonVar.Project).ShouldPass().Out()
-			Expect(appListOutput).To(ContainSubstring(appName))
-			actualCompListJSON := helper.Cmd("odo", "app", "list", "-o", "json", "--project", commonVar.Project).ShouldPass().Out()
-			valuesList := gjson.GetMany(actualCompListJSON, "kind", "items.#.metadata.name", "items.#.metadata.namespace")
-			expectedList := []string{"List", "app", commonVar.Project}
-			Expect(helper.GjsonMatcher(valuesList, expectedList)).To(Equal(true))
-
-			helper.Cmd("odo", "app", "describe", appName, "--project", commonVar.Project).ShouldPass()
-			actualDesAppJSON := helper.Cmd("odo", "app", "describe", appName, "--project", commonVar.Project, "-o", "json").ShouldPass().Out()
-			valuesDes := gjson.GetMany(actualDesAppJSON, "kind", "metadata.name", "metadata.namespace")
-			expectedDes := []string{"Application", appName, commonVar.Project}
-			Expect(helper.GjsonMatcher(valuesDes, expectedDes)).To(Equal(true))
-
-			helper.Cmd("odo", "app", "delete", appName, "--project", commonVar.Project, "-f").ShouldPass()
 		})
 
-	})
+		When("odo push is executed", func() {
+
+			BeforeEach(func() {
+				helper.Cmd("odo", "push", "--context", commonVar.Context).ShouldPass()
+			})
+
+			It("should fail describing app without app parameter", func() {
+				helper.Cmd("odo", "app", "describe", "--project", commonVar.Project).ShouldFail()
+			})
+
+			It("should fail deleting an app without app parameter", func() {
+				helper.Cmd("odo", "app", "delete", "-f", "--project", commonVar.Project).ShouldFail()
+			})
+
+			It("should list apps", func() {
+				appListOutput := helper.Cmd("odo", "app", "list", "--project", commonVar.Project).ShouldPass().Out()
+				Expect(appListOutput).To(ContainSubstring(appName))
+			})
+
+			It("should list apps in JSON format", func() {
+				actualCompListJSON := helper.Cmd("odo", "app", "list", "-o", "json", "--project", commonVar.Project).ShouldPass().Out()
+				valuesList := gjson.GetMany(actualCompListJSON, "kind", "items.#.metadata.name", "items.#.metadata.namespace")
+				expectedList := []string{"List", "app", commonVar.Project}
+				Expect(helper.GjsonMatcher(valuesList, expectedList)).To(Equal(true))
+			})
+
+			It("should describe specific app", func() {
+				helper.Cmd("odo", "app", "describe", appName, "--project", commonVar.Project).ShouldPass()
+			})
+
+			It("should describe specific app in JSON format", func() {
+				actualDesAppJSON := helper.Cmd("odo", "app", "describe", appName, "--project", commonVar.Project, "-o", "json").ShouldPass().Out()
+				valuesDes := gjson.GetMany(actualDesAppJSON, "kind", "metadata.name", "metadata.namespace")
+				expectedDes := []string{"Application", appName, commonVar.Project}
+				Expect(helper.GjsonMatcher(valuesDes, expectedDes)).To(Equal(true))
+			})
+
+			When("changing to context directory", func() {
+
+				BeforeEach(func() {
```

Comment:
We have defined context in `BeforeEach` https://github.com/openshift/odo/pull/4806/files#diff-ae5d2699e830fc0052b86751a2d5816e2f6381a8d27e71c28bb56e8c44651b26R17 and then modifying the content of it now. AFAIK in such scenario we go for `JustBeforeEach`. Correct me if I am wrong.

Reply:
This is not what I understand about `JustBeforeEach`.

`BeforeEach`es are exectued one after the other, for example:

```
When("...", func() {
  BeforeEach(func() {
     fmt.Printf("Executed #1")
  }

  When("...", func() {
    BeforeEach(func() {
      fmt.Printf("Executed #2")
    }
  })
})
```

When using `JustBeforeEach`, they will be executed after all `BeforeEach`es:

```
When("...", func() {
  BeforeEach(func() {
     fmt.Printf("Executed #1")
  }

  JustBeforeEach(func() {
     fmt.Printf("Executed #3")
  }

  When("...", func() {
    BeforeEach(func() {
      fmt.Printf("Executed #2")
    }
  })
})
```